src/providers/info.rs

```diff
@@ -138,6 +145,87 @@ impl InfoProvider for DefaultInfoProvider<'_> {
 }
 
 impl DefaultInfoProvider<'_> {
+    pub fn new<'p>(context: &'p Context) -> DefaultInfoProvider<'p> {
+        DefaultInfoProvider {
+            client: context.client(),
+            cache: context.cache(),
+            safe_cache: HashMap::new(),
+            token_cache: HashMap::new(),
+        }
+    }
+
+    fn cached<T>(
+        &mut self,
+        local_cache: impl Fn(&mut Self) -> &mut HashMap<String, Option<T>>,
+        generator: impl Fn(&mut Self, &String) -> ApiResult<Option<T>>,
+        key: impl Into<String>,
+    ) -> ApiResult<T>
+    where
+        T: Clone + DeserializeOwned,
+    {
+        let key = key.into();
+        match local_cache(self).get(&key) {
+            Some(value) => value
+                .clone()
+                .ok_or(api_error!("Cached value not available")),
+            None => {
+                let value: Option<T> = generator(self, &key)?;
+                local_cache(self).insert(key, value.clone());
+                value.ok_or(api_error!("Could not generate value"))
+            }
+        }
+    }
+
+    fn load_safe_info(&mut self, safe: &String) -> ApiResult<Option<SafeInfo>> {
+        let url = format!("{}/v1/safes/{}/", base_transaction_service_url(), safe);
+        let data: String = self
+            .cache
+            .request_cached(self.client, &url, info_cache_duration())?;
+        Ok(serde_json::from_str(&data).unwrap_or(None))
+    }
+
+    fn populate_token_cache(&mut self) -> ApiResult<()> {
+        let url = format!("{}/v1/tokens/?limit=10000", base_transaction_service_url());
+        let response = self.client.get(&url).send()?;
+        let data: Page<TokenInfo> = response.json()?;
+        for token in data.results.iter() {
+            self.cache.create(
+                &format!("dip_ti_{}", token.address),
+                &serde_json::to_string(&token)?,
+                info_cache_duration(),
+            )
+        }
+        Ok(())
+    }
+
+    fn check_token_cache(&mut self) -> ApiResult<()> {
+        if self.cache.fetch("dip_tcl").is_some() {
+            // Cache is still up to data
+            return Ok(());
+        }
+        let result = self.populate_token_cache();
+        // If error we use a shorter cache timeout (we do not want to DoS our service in case of an error)
+        self.cache.create(
+            "dip_tcl",
+            "",
+            if result.is_ok() {
+                info_cache_duration()
+            } else {
+                10000
```

Comment:
There is this hardcoded timeout that I am not sure what it would be. This would be the error timeout ? Maybe we can merge the other PR before continuing with this one. https://github.com/gnosis/safe-client-gateway/pull/282

Reply:
yes this is the error timeout ... can be hardcoded for a staging version but as you mentioned once #282 We should update this (maybe add a TODO commennt?)

Reply:
I've merged the error timeout issue, so I will just add it. We can adjust the values according to #278